middle_end/flambda2/cmx/flambda_cmx.ml

```diff
@@ -191,78 +190,90 @@ let compute_reachable_names_and_code ~module_symbol typing_env code =
-    in
-    let final_typing_env =
-      TE.Serializable.create final_typing_env ~reachable_names
-    in
-    (* We need to re-export offsets for everything reachable from the cmx file;
-       closure_vars/ids can be reachable from the typing env, and the exported
-       code.
-
-       In the case of the exported code, we already have offsets for all
-       closure_ids/vars reachable from the code of the current compilation unit,
-       but since we also re-export code from other compilation units, we need to
-       take those into account. *)
-    (* CR gbury: it might be more efficient to not compute the free names for
-       all exported code, but fold over the exported code to avoid allocating
-       some free_names *)
-    let free_names_of_all_code = EC.free_names all_code in
-    let closure_elts_used_in_typing_env =
-      TE.Serializable.free_closure_ids_and_closure_vars final_typing_env
+    let create_typing_env reachable_names =
+      TE.Serializable.create typing_env ~reachable_names
     in
-    let exported_offsets =
-      exported_offsets
-      |> Closure_offsets.reexport_closure_ids
-           (Name_occurrences.closure_ids free_names_of_all_code)
-      |> Closure_offsets.reexport_closure_vars
-           (Name_occurrences.closure_vars free_names_of_all_code)
-      |> Closure_offsets.reexport_closure_ids
-           (Name_occurrences.closure_ids closure_elts_used_in_typing_env)
-      |> Closure_offsets.reexport_closure_vars
-           (Name_occurrences.closure_vars closure_elts_used_in_typing_env)
+    let free_names_of_name name =
+      Option.map T.free_names
+        (TE.Pre_serializable.find_or_missing typing_env name)
     in
-    Some
-      (Flambda_cmx_format.create ~final_typing_env ~all_code ~exported_offsets
-         ~used_closure_vars)
+    prepare_cmx ~module_symbol create_typing_env ~free_names_of_name
+      ~used_closure_vars ~canonicalise ~exported_offsets all_code
 
-let prepare_cmx_from_approx ~approxs ~exported_offsets ~used_closure_vars
-    all_code =
+let prepare_cmx_from_approx ~approxs ~module_symbol ~exported_offsets
+    ~used_closure_vars all_code =
   if Flambda_features.opaque ()
   then None
   else
-    (* CR keryan : for now all content of loaded .cmx are reexported. We may
-       want to remove unreachable code as does [prepare_cmx_file_contents] at
-       some point *)
-    let final_typing_env =
+    let typing_env =
       TE.Serializable.create_from_closure_conversion_approx approxs
     in
-    let free_names_of_all_code = EC.free_names all_code in
-    let exported_offsets =
-      exported_offsets
-      |> Closure_offsets.reexport_closure_ids
-           (Name_occurrences.closure_ids free_names_of_all_code)
-      |> Closure_offsets.reexport_closure_vars
-           (Name_occurrences.closure_vars free_names_of_all_code)
+    let create_typing_env _ = typing_env in
+    let free_names_of_name name =
+      match Name.must_be_symbol_opt name with
+      | None -> None
```

Comment:
I'd rather have a fatal error here. For now it's not supposed to happen, as neither code nor the symbol approximations should contain variables, and if we later allow variables in approximations then we need to correctly account for the associated free names.